### PR TITLE
2.x Authentication

### PR DESCRIPTION
**Pull request by e0ipso**
This PR depends on https://github.com/RESTful-Drupal/restful/pull/441.

Test the different authentication providers:
- [x] Cokie authentication.
- [x] Basic authentication.
- [x] No authentication.
- [x] Login endpoint.

### CODE REVIEW

**Comment by e0ipso** [comment]
I am leaving token auth out of this PR. I'll open a new one to migrate it.

**Comment by amitaibu** [comment]
Ready for review?

**Comment by e0ipso** [comment]
yes!

File: src/Http/RequestInterface.php [hunk 1]
```diff
@@ -243,4 +243,28 @@ public function isSecure();
...
+  /**
+   * Is via router.
+   *
+   * @return bool
+   *   TRUE if the request was generated via the Drupal's menu router.
+   */
+  public function isViaRouter();
+
+  /**
+   * Marks the request as being via router.
```
Comment (amitaibu) [inline src/Http/RequestInterface.php]: => `via Drupal's menu router.`

File: src/Plugin/resource/LoginCookie__1_0.php [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  public function loginAndRespondWithCookie() {
+    // Login the user.
+    $account = $this->getAccount();
+    $this->loginUser($account);
+
+    $user_resource = restful()
+      ->getResourceManager()
+      ->getPlugin('users:1.0');
+
+    $output = $user_resource ? $user_resource->view($account->uid) : array();
```
Comment (amitaibu) [inline src/Plugin/resource/LoginCookie__1_0.php:91]: `// User resource may be disabled.`

File: src/Plugin/resource/LoginCookie__1_0.php [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    $output += restful_csrf_session_token();
+    return $output;
+  }
+
+  /**
+   * Log the user.
```
Comment (amitaibu) [inline src/Plugin/resource/LoginCookie__1_0.php]: => `Log in the user.`

File: src/Plugin/resource/Resource.php [hunk 4]
```diff
@@ -410,4 +421,33 @@ protected function processPublicFields(array $field_definitions) {
     return $field_definitions;
   }
 
+  /**
+   * Initializes the authentication manager and adds the appropriate providers.
+   */
+  protected function initAuthenticationManager() {
+    // If there is authentication enabled, create the authentication manager.
```
Comment (amitaibu) [inline src/Plugin/resource/Resource.php]: Can you explain what you mean by `If there is authentication enabled`?

File: src/Plugin/resource/Resource.php [hunk 5]
```diff
@@ -410,4 +421,33 @@ protected function processPublicFields(array $field_definitions) {
...
+    $this->authenticationManager->setIsOptional($authentication_optional);
+    if (empty($authentication_types)) {
+      if (empty($authentication_optional)) {
+        // Fail early, fail good.
+        throw new UnauthorizedException('There are no authentication providers and authentication is not optional.');
+      }
+      return;
+    }
+    if ($authentication_types === TRUE) {
+      // All all the available authentication providers to the manager.
```
Comment (amitaibu) [inline src/Plugin/resource/Resource.php]: `// All` => `// Add ...`

File: src/Plugin/resource/Users__1_0.php [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  protected function publicFields() {
+    $public_fields = array();
+    $public_fields['id'] = array(
+      'property' => 'uid',
+      'methods' => array(RequestInterface::METHOD_GET),
+    );
+    $public_fields['mail'] = array(
```
Comment (amitaibu) [inline src/Plugin/resource/Users__1_0.php:40]: Maybe worth adding a comment that the `mail` will be shown only to the own user or privileged users.

File: src/Plugin/resource/Users__1_0.php [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+      'property' => 'mail',
+    );
+    $public_fields['self'] = array(
```
Comment (amitaibu) [inline src/Plugin/resource/Users__1_0.php]: why not inherit from `parent::publicFields()`?

**Comment by amitaibu** [comment]
Minor stuff - things look good :+1:

**Comment by e0ipso** [comment]
Fixed feedback and merging!

**Pull request merged**